Foreign Library Stanza

ropwareJB opened the pull request:
- [Y ] Did you add a new feature to `hpack`?
  - [Y] Did you add an acceptance test for that new feature to `test/EndToEndSpec.hs`?

Addresses https://github.com/sol/hpack/issues/258.

Adds the following syntax to support [foreign / shared libraries](https://cabal.readthedocs.io/en/3.4/cabal-package.html#foreign-libraries) so we dont need to use the verbatim block anymore:

```yaml
foreign-library:
  type: native-shared
  source-dirs: driver
  options:
    - standalone
  dependencies:
    - hpack
```

ropwareJB committed:
Support foreign library stanza

mpilgrem commented:
I do not know why the CI is failing for Windows/'system' GHC (being GHC 9.6.2). I logged https://gitlab.haskell.org/ghc/ghc/-/issues/23533 as it seems to be GHC-related.

Locally, I can build and test (`stack test`) the Hpack respository at commit 36c44de831ebaf3962763ffa2c34a424b2a8650e with Stack configuration:
~~~yaml
resolver: nightly-2023-06-18 # GHC 9.4.5
compiler: ghc-9.6.2

extra-deps:
# GHC 9.6.2 boot libaries
- Cabal-3.10.1.0
- Cabal-syntax-3.10.1.0
- directory-1.3.8.1
- time-1.12.2
- process-1.6.17.0
- Win32-2.13.3.0
# Required because the Stackage snapshot versions do not support GHC 9.6.2
- semigroupoids-6.0.0.1
- bifunctors-5.6.1
~~~

sol commented:
Thanks @mpilgrem.

mpilgrem commented:
I have rebased on the `main` branch and squashed the existing commits making up the pull request.

mpilgrem commented:
@sol, I think this is good to merge. Do you want to review?

sol commented:
> @sol, I think this is good to merge. Do you want to review?

@mpilgrem Hold back first, I'm gonna look at it sometime soon.

sol commented:
@mpilgrem I rebased again.  But commits don't seem to be squashed yet.

mpilgrem commented:
@sol, I've squashed but I was not sure if your review was complete subject to that?

sol reviewed (commented):
~First, despite what I was saying earlier, I want to raise the questions "Do we really want this?" and "Is this within the scope of Hpack?"~

~For Linux I don't know, I think you usually would go the other way around and use foreign imports.  Are there any plausible use cases? If it is just for experimentation, then it would probably be outside the scope of Hpack.~

~For Windows however, the GHC documentation provides a VBA example.  So if people wanna move business logic from Excel to Haskell then I can sympathize with that.~

~Maybe that's enough as a justification?~

EDIT: I think this can actually be useful in combination with Python and `ctypes`, so let's go for it.

Regarding the implementation, I think `type` is a no-brainer, just remove it.

`options` on the other hand is more complicated.  From what I understand, if we drop `options` then you would unconditionally need to generate

```cabal
if os(Windows)
  options: standalone
```

While this is in line with "Don't require the user to state the obvious" I'm still puzzled how we would achieve "Give the user 100% control when needed" here? I think a user will not be able to override this with `verbatim`.  But maybe just drop it anyway and address the 100% control thing once somebody asks for it?

BTW: ~I kind of think that if we decide to include this feature then we should probably include a complete example in the README, similar to what the GHC documentation contains, so that a user does not need to traverse three different manuals to make sense of this.~  EDIT: Even simple examples are still more complicated than I would hope for, so I'm not gonna insist on this just yet.

sol commented:
@mpilgrem  I'm probably busy tomorrow, but gonna try to look at it again on Monday.

sol commented on src/Hpack/Config.hs:
```diff
@@ -1241,25 +1313,28 @@ type GlobalOptions = CommonOptions AsmSources CSources CxxSources JsSources Empt
     globalOptions = g {commonOptionsVerbatim = Nothing}
 
-    executableNames = maybe [] Map.keys executableMap
+    componentNames = maybe [] Map.keys executableMap ++ maybe [] Map.keys foreignLibraryMap
```
Specifically I think the old is just fine, we don't need to update this.

mpilgrem replied:
I will revert the removal of `executableNames`.

sol commented on src/Hpack/Config.hs:
```diff
@@ -611,6 +640,8 @@ data PackageConfig_ library executable = PackageConfig {
 , packageConfigLibrary :: Maybe library
 , packageConfigInternalLibraries :: Maybe (Map String library)
+, packageConfigForeignLibraries :: Maybe (Map String foreignLib)
+, packageConfigForeignLibrary :: Maybe foreignLib
```
I guess it is actually a good idea, so let's add the test.

mpilgrem replied:
I will add a test so that both `foreign-libraries` and `foreign-library` are tested.

mpilgrem commented:
@ropwareJB, @sol, I'm going to pay this PR some attention. I'll start by rebasing on `main`.

mpilgrem committed:
Relocate foreign-libraries in README.md

This weakens a possible implication that libraries provided by a Haskell package and foreign-libraries are somehow related.

mpilgrem committed:
Add foreign-library to README.md

mpilgrem committed:
In README.md, conform foreign-libraries to executables

mpilgrem commented on README.md:
```diff
@@ -328,6 +329,17 @@ This is done to allow compatibility with a wider range of `Cabal` versions.
+| `type` | . | | |
+| `lib-version-info` | . | | |
+| `options` | . | | |
+| `mod-def-file` | . | | |
```
I'll remove all of `type`, `options` and `mod-def-file`.

mpilgrem committed:
Foreign-library: Fix README.md for other-modules and generated-other-modules

mpilgrem commented on README.md:
```diff
@@ -328,6 +329,17 @@ This is done to allow compatibility with a wider range of `Cabal` versions.
+| `lib-version-info` | . | | |
+| `options` | . | | |
+| `mod-def-file` | . | | |
+| `other-modules` | · | All modules in `source-dirs` less `main` less any modules mentioned in `when` | |
```
I'll remove the reference to a `main` field.

mpilgrem commented on README.md:
```diff
@@ -328,6 +329,17 @@ This is done to allow compatibility with a wider range of `Cabal` versions.
+| `options` | . | | |
+| `mod-def-file` | . | | |
+| `other-modules` | · | All modules in `source-dirs` less `main` less any modules mentioned in `when` | |
+| `generated-other-modules` | | | Added to `other-modules` and `autogen-modules`. Since `0.23.0`.
```
I'll remove the `since` annotation.

mpilgrem committed:
Revert removal of executableNames

mpilgrem commented on README.md:
```diff
@@ -328,6 +329,17 @@ This is done to allow compatibility with a wider range of `Cabal` versions.
+
+| Hpack | Cabal | Default | Notes |
+| --- | --- | --- | --- |
+| `type` | . | | |
```
I'll remove the field.

mpilgrem commented on README.md:
```diff
@@ -328,6 +329,17 @@ This is done to allow compatibility with a wider range of `Cabal` versions.
+| --- | --- | --- | --- |
+| `type` | . | | |
+| `lib-version-info` | . | | |
+| `options` | . | | |
```
I'll remove the field.

mpilgrem committed:
foreign-libraries: Revise test and add test

mpilgrem commented on test/EndToEndSpec.hs:
```diff
@@ -2099,6 +2112,17 @@ library #{name}
+    content = [i|
+foreign-library #{name}
+  other-modules:
+      Paths_my_package
```
I will make `foreign-libraries:` not respect the legacy Paths_ module handling.

mpilgrem commented on test/EndToEndSpec.hs:
```diff
@@ -1696,6 +1696,19 @@ spec = around_ (inTempDirectoryNamed "my-package") $ do
+        options:
+            standalone
+        |])
+
```
I will include tests for both `foreign-libraries` and `foreign-library`.

mpilgrem committed:
Foreign-libraries: remove type, options and mod-def-file keys

Also does not respect legacy Paths_ module handling.

mpilgrem commented:
@ropwareJB, @sol, I have a couple of questions:

On Linux only, Cabal supports a [`lib-version-linux` field](https://hackage-content.haskell.org/package/Cabal-syntax-3.16.1.0/docs/Distribution-Types-ForeignLib.html#v:foreignLibVersion) (API version) and a [`lib-version-info` field](https://cabal.readthedocs.io/en/stable/cabal-package-description-file.html#pkg-field-foreign-library-lib-version-info) (a GNU Libtool `current[:revision[:age]]` triple).

The former takes precedence over the latter, if both are specified:
* https://hackage-content.haskell.org/package/Cabal-syntax-3.16.1.0/docs/Distribution-Types-ForeignLib.html#v:foreignLibVersion

Q1. The original pull request assumes that Hpack will implement `lib-version-info` only. Is that the intention?

Q2. With this pull request, Hpack itself does not check that the supplied `lib-version-info` string is a valid Libtool triple. Is that how Hpack should behave?

mpilgrem commented on src/Hpack/Config.hs:
```diff
@@ -194,6 +197,7 @@ renameDependencies old new sect@Section{..} = sect {sectionDependencies = (Depen
 packageDependencies :: Package -> [(String, DependencyInfo)]
 packageDependencies Package{..} = nub . sortBy (comparing (lexicographically . fst)) $
      (concatMap deps packageExecutables)
+  ++ (concatMap deps packageForeignLibraries)
```
Unrelated to this pull request.

mpilgrem commented on src/Hpack/Config.hs:
```diff
@@ -176,6 +178,7 @@ package name version = Package {
 renamePackage :: String -> Package -> Package
 renamePackage name p@Package{..} = p {
     packageName = name
+  , packageForeignLibraries = fmap (renameDependencies packageName name) packageForeignLibraries
```
Unrelated to this pull request.
